libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+                                                                                  typename Traits::difference_type,
+                                                                                  typename Traits::pointer,
+                                                                                  typename Traits::reference>;
+
+    /************************
+     * select_iterator_base *
+     ************************/
+
+    namespace detail
+    {
+        template <class Iterator>
+        using iterator_category_t = typename std::iterator_traits<Iterator>::iterator_category;
+    
+        template <class Iterator>
+        using is_random_access_iterator = std::is_same<iterator_category_t<Iterator>, std::random_access_iterator_tag>;
+    }
+
+
+    template <class Iterator>
+    struct select_iterator_base
+    {
+        template <class I, class T, class D = std::ptrdiff_t, class P = T*, class R = T&>
+        using type = std::conditional_t<detail::is_random_access_iterator<Iterator>::value,
+                                        xrandom_access_iterator_base<I, T, D, P, R>,
+                                        xbidirectional_iterator_base<I, T, D, P, R>>;
+
+        template <class I, class T>
+        using from_traits = std::conditional_t<detail::is_random_access_iterator<Iterator>::value,
+                                               xrandom_access_iterator_base_from_traits<I, T>,
+                                               xbidirectional_iterator_base_from_traits<I, T>>;
+    };
+
+    template <class Iterator, class R = void>
+    using enable_random_access_iterator = std::enable_if_t<detail::is_random_access_iterator<Iterator>::value, R>;
+
+    /*******************
+     * filter_iterator *
+     *******************/
+
+    template <class Predicate, class Iterator>
+    class filter_iterator;
+
+    template <class Predicate, class Iterator>
+    struct select_filter_iterator_base
+    {
+        using type = typename select_iterator_base<Iterator>::from_traits<
+            filter_iterator<Predicate, Iterator>,
+            std::iterator_traits<Iterator>
+        >;
+    };
+
+    template <class Predicate, class Iterator>
+    using select_filter_iterator_base_t = typename select_filter_iterator_base<Predicate, Iterator>::type;
+
+    template <class Predicate, class Iterator>
+    class filter_iterator : public select_filter_iterator_base_t<Predicate, Iterator>
+    {
+    public:
+
+        using self_type = filter_iterator<Predicate, Iterator>;
+        using base_type = select_filter_iterator_base_t<Predicate, Iterator>;
+
+        using reference = typename base_type::reference;
+        using pointer = typename base_type::pointer;
+        using difference_type = typename base_type::difference_type;
+        using iterator_category = typename base_type::iterator_category;
+
+        filter_iterator() = default;
```

Comment:
Is this iterator usable in any way when initialized by default? Other than through re-assigning it?

Reply:
It is not, but iterators must be default constructible according to the standard.